### PR TITLE
feat: iOS ListView

### PR DESCRIPTION
**Pull request by lucasaraujo**
### Related Issues

#210 

### Checklist

Please, check if these important points are met using `[x]`:

- [x] I read the [PR Guide] and followed the process outlined there for submitting this PR.
- [x] I avoided _breaking changes_ by not changing public APIs that people rely on. 
- [x] I am willing to follow-up on review comments in a timely manner.

[PR Guide]: https://github.com/ZupIT/beagle/blob/master/doc/contributing/pull_requests.md

### CODE REVIEW

File: iOS/Example/BeagleDemo/BeagleDemo/Screens/ListViewScreen.swift [hunk 1]
```diff
@@ -30,43 +30,70 @@ struct ListViewScreen: DeeplinkScreen {
...
-        }
-    }
+    )
+    
+    var listView = ListView(
+        context: Context(
+            id: "initialContext",
+            value: ""
+        ),
+        onInit: [SendRequest(
```
Comment (botbeaglezup) [inline iOS/Example/BeagleDemo/BeagleDemo/Screens/ListViewScreen.swift:60]: <table data-meta="generated_by_ios">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/8912c9e5b8367e431f8a1d8b85b9c378ac7cf2de/iOS/Example/BeagleDemo/BeagleDemo/Screens/ListViewScreen.swift#L60"></span>Multiline literals should have their surrounding brackets in a new line.<br />
<code>multiline_literal_brackets</code> <code>ListViewScreen.swift:60</code></td>
    </tr>
  </tbody>
</table>

File: iOS/Example/BeagleDemo/BeagleDemo/Screens/ListViewScreen.swift [hunk 2]
```diff
@@ -30,43 +30,70 @@ struct ListViewScreen: DeeplinkScreen {
...
+            url: "https://run.mocky.io/v3/28f3ad29-07ed-44b3-af95-a6986bfd8565",
+            method: .get,
+            onSuccess: [
+                SetContext(
+                    contextId: "initialContext",
+                    value: "@{onSuccess.data.genres}"
+                )
+            ]
+        )],
```
Comment (botbeaglezup) [inline iOS/Example/BeagleDemo/BeagleDemo/Screens/ListViewScreen.swift:69]: <table data-meta="generated_by_ios">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/8912c9e5b8367e431f8a1d8b85b9c378ac7cf2de/iOS/Example/BeagleDemo/BeagleDemo/Screens/ListViewScreen.swift#L69"></span>Multiline literals should have their surrounding brackets in a new line.<br />
<code>multiline_literal_brackets</code> <code>ListViewScreen.swift:69</code></td>
    </tr>
  </tbody>
</table>

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/ZupIT/beagle/blob/590ef8f520ad7934de7ece7a0440e2210b5f3f49/iOS/Sources/Beagle/Sources/Renderer/BeagleRenderer.swift#L140">iOS/Sources/Beagle/Sources/Renderer/BeagleRenderer.swift#L140</a> - TODOs should be resolved (should we make <code>observeMany</code> t…).<br />
<code>todo</code> <code>BeagleRenderer.swift:140</code></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_ios">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by lucasfonsecazup — COMMENTED** [review]
Some non-impeding comments, in general, looks great.

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/ZupIT/beagle/blob/3795ca9e8018bf30926efab6d34752d6b03f96c1/iOS/Sources/Beagle/Sources/Renderer/BeagleRenderer.swift#L132">iOS/Sources/Beagle/Sources/Renderer/BeagleRenderer.swift#L132</a> - TODOs should be resolved (should we make <code>observeMany</code> t…).<br />
<code>todo</code> <code>BeagleRenderer.swift:132</code></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_ios">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: iOS/Sources/Beagle/Sources/Components/ListView+Renderable/ListViewUIComponent.swift [hunk 3]
```diff
@@ -17,123 +17,297 @@
...
+    }
+    
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        executeOnScrollEndIfNeededAfterLayout()
+    }
+    
+    // MARK: - Sizing
+    
+    override func sizeThatFits(_ size: CGSize) -> CGSize {
```
[the remark this answers is not in the document]
Comment (lucasaraujo) [reply iOS/Sources/Beagle/Sources/Components/ListView+Renderable/ListViewUIComponent.swift:122]: Documentation added.

File: iOS/Sources/Beagle/Sources/Components/Layout+Renderable/Style/Tests/ContainerTests.swift [hunk 4]
```diff
@@ -119,12 +119,17 @@ class BeagleControllerSpy: BeagleController {
     
     var expectation: XCTestExpectation?
     
-    func addBinding(_ update: @escaping () -> Void) { }
-    func execute(action: RawAction, sender: Any) { }
-    
     private(set) var didCalledExecute = false
     private(set) var lastImplicitContext: Context?
     
+    func addOnInit(_ onInit: [RawAction], in view: UIView) {
```
[the remark this answers is not in the document]
Comment (lucasaraujo) [reply iOS/Sources/Beagle/Sources/Components/Layout+Renderable/Style/Tests/ContainerTests.swift:125]: I agree, but we don't have a folder to the tests, so I'm not sure where the file should be placed. Any idea?

[the remark this answers is not in the document]
Comment (lucasfonsecazup) [reply iOS/Sources/Beagle/Sources/Components/Layout+Renderable/Style/Tests/ContainerTests.swift:125]: There are some files in `beagle / iOS / Sources / Beagle / BeagleTests /`

File: iOS/Sources/Beagle/Sources/Components/ListView+Renderable/ListViewCell.swift [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    
+    private var bindings = [() -> Void]()
+    private var onInits = [(actions: [RawAction], view: UIView)]()
+    private var pendingActions = Set<UUID>()
+    
+    private var templateContainer: TemplateContainer?
+    private weak var listView: ListViewUIComponent?
+    
+    var templateIntrinsicSize: CGSize {
+        guard let template = templateContainer else { return .zero }
```
Comment (lucasfonsecazup) [inline iOS/Sources/Beagle/Sources/Components/ListView+Renderable/ListViewCell.swift:34]: You don't need that `guard let`, we can just unwrap the value:

> return templateContainer?.yoga.calculateLayout(
>            with: CGSize(width: CGFloat.nan, height: CGFloat.nan)
>        ) ?? .zero

File: iOS/Sources/Beagle/Sources/Components/ListView+Renderable/ListViewUIComponent.swift [hunk 6]
```diff
@@ -17,123 +17,309 @@
...
 import BeagleSchema
 
-// MARK: - Model
-extension ListViewUIComponent {
-    struct Model {
-        let component: ListView
-        var componentViews: [(view: UIView, size: CGSize)]
+private struct CellsContextManager {
+    
+    private var orphanCells = [Int: ListViewCell]()
```
Comment (lucasfonsecazup) [inline iOS/Sources/Beagle/Sources/Components/ListView+Renderable/ListViewUIComponent.swift:22]: I loved this name <3

File: iOS/Sources/Beagle/Sources/Components/ListView+Renderable/ListViewUIComponent.swift [hunk 7]
```diff
@@ -17,123 +17,309 @@
...
+                return value
+            case .empty, .bool, .array, .dictionary, .expression:
+                break
+            }
+        }
+        return nil
     }
     
+    private func hashFor(item: DynamicObject, withKey key: String?) -> Int {
+        if let key = key {
```
Comment (lucasfonsecazup) [inline iOS/Sources/Beagle/Sources/Components/ListView+Renderable/ListViewUIComponent.swift:282]: We don't need this `if let` we can unwrap it with a default value: 
> key?.hashValue ?? item.hashValue

File: iOS/Sources/Beagle/Sources/Components/ListView+Renderable/Tests/ListViewTests.swift [hunk 8]
```diff
@@ -23,21 +23,204 @@ final class ListViewTests: XCTestCase {
...
+        )
+    }
+    
+    private lazy var controller = BeagleScreenViewController(ComponentDummy())
+    
+    private func renderListView(_ listComponent: ListView) -> UIView {
+        let host = ComponentHostController(listComponent, renderer: controller.renderer)
+        return host.view
+    }
+
```
Comment (lucasfonsecazup) [inline iOS/Sources/Beagle/Sources/Components/ListView+Renderable/Tests/ListViewTests.swift:56]: Nice tests, but according to convention we are using the `Given / When / Then` comments -> https://github.com/ZupIT/beagle/wiki/Swift-Style-Guide#tests

**Review by lucasfonsecazup — COMMENTED** [review]
:)

File: iOS/Sources/Beagle/Sources/Renderer/BeagleScreenViewController.swift [hunk 9]
```diff
@@ -48,6 +52,8 @@ public class BeagleScreenViewController: BeagleController {
     
     private var bindings: [() -> Void] = []
     
+    private var onInit: [(UIView, [RawAction])] = []
+    
     private var navigationControllerId: String?
     
     // TODO: This workaround should be removed in BeagleView future implementation
```
Comment (botbeaglezup) [inline iOS/Sources/Beagle/Sources/Renderer/BeagleScreenViewController.swift:59]: <table data-meta="generated_by_ios">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/637fd2f7333326180eb8f9835d81295e32e74887/iOS/Sources/Beagle/Sources/Renderer/BeagleScreenViewController.swift#L59"></span>TODOs should be resolved (This workaround should be remo…).<br />
<code>todo</code> <code>BeagleScreenViewController.swift:59</code></td>
    </tr>
  </tbody>
</table>

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/ZupIT/beagle/blob/33b20cf0e1e708a5125aaf3bede65ad03c739b82/android/automated-tests/src/main/AndroidManifest.xml#L8">android/automated-tests/src/main/AndroidManifest.xml#L8</a> - On SDK version 23 and up, your app data will be automatically backed up and restored on app install. Consider adding the attribute <code>android:fullBackupContent</code> to specify an <code>@xml</code> resource which configures which files to backup. More info: https://developer.android.com/training/backup/autosyncapi.html</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/ZupIT/beagle/blob/33b20cf0e1e708a5125aaf3bede65ad03c739b82/android/automated-tests/src/main/AndroidManifest.xml#L16">android/automated-tests/src/main/AndroidManifest.xml#L16</a> - Attribute <code>usesCleartextTraffic</code> is only used in API level 23 and higher (current min is 19)</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/ZupIT/beagle/blob/33b20cf0e1e708a5125aaf3bede65ad03c739b82/android/beagle/src/main/java/br/com/zup/beagle/android/utils/ToolbarManager.kt#L224">android/beagle/src/main/java/br/com/zup/beagle/android/utils/ToolbarManager.kt#L224</a> - Prefer “<code>SHOW_AS_ACTION_IF_ROOM</code>” instead of “<code>SHOW_AS_ACTION_ALWAYS</code>”</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_android">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/ZupIT/beagle/blob/14d79ae497e8368e1f9fcfad884aa8294b0ef8ec/iOS/Sources/Beagle/Sources/Renderer/BeagleRenderer.swift#L145">iOS/Sources/Beagle/Sources/Renderer/BeagleRenderer.swift#L145</a> - TODOs should be resolved (should we make <code>observeMany</code> t…).<br />
<code>todo</code> <code>BeagleRenderer.swift:145</code></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_ios">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by dantes-git — APPROVED** [review]

**Review by lucasfonsecazup — APPROVED** [review]

File: iOS/Sources/Beagle/Sources/Components/Layout+Renderable/Style/Tests/ContainerTests.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucasaraujo) [reply iOS/Sources/Beagle/Sources/Components/Layout+Renderable/Style/Tests/ContainerTests.swift:152]: done

File: iOS/Schema/Sources/Action/Types/SendRequest.swift [hunk 10]
```diff
@@ -33,7 +33,7 @@ public struct SendRequest: RawAction, AutoInitiableAndDecodable {
     public let headers: Expression<[String: String]>?
     public let onSuccess: [RawAction]?
     public let onError: [RawAction]?
-    public let onFinish: [RawAction]?
+    public var onFinish: [RawAction]?
```
[the remark this answers is not in the document]
Comment (lucasaraujo) [reply iOS/Schema/Sources/Action/Types/SendRequest.swift:36]: the AsyncAction protocol requires the property get method

**Pull request merged**